File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9D.scala
package TextbookTrack.Chapter09

object BA9D {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}

  final case class Edge(nodeTo: Int, startIx: Int, length: Int)

  class SuffixTree(val text: String) {
    private val size: Int = text.length
    private val nodeIdGenerator: Iterator[Int] = Iterator.from(0)

    val root: Int = nodeIdGenerator.next()
    val adjacencyList: Map[Int, List[Edge]] = buildSuffixTree()

    def edges(): Iterator[String] =
      adjacencyList
        .valuesIterator
        .flatMap(_.map{ case Edge(_, startIx, length) => text.slice(startIx, startIx + length) })

    private def buildSuffixTree(): Map[Int, List[Edge]] = {
      val adjacencyList: MutableMap[Int, List[Edge]] = MutableMap()
      (0 until size).foreach(addSuffixToTree(adjacencyList, _))
      adjacencyList.toMap
    }

    private def addSuffixToTree(adjacencyList: MutableMap[Int, List[Edge]], suffixStart: Int): Unit = {
      val (node, optionalEdge, edgeStartIx, nonMatchingIx): (Int, Option[Edge], Int, Int) =
        findLastMatchingNode(adjacencyList, suffixStart)
      val newLeaf: Int = nodeIdGenerator.next()
      optionalEdge match {
        case None =>
          adjacencyList
            .update(node, Edge(newLeaf, edgeStartIx, size - edgeStartIx) :: adjacencyList.getOrElse(node, Nil))
        case Some(edge @ Edge(nodeTo, startIx, length)) =>
          val newNode: Int = nodeIdGenerator.next()
          val splitLength1: Int = nonMatchingIx - edgeStartIx
          val splitLength2: Int = length - splitLength1
          adjacencyList(node) = adjacencyList(node).filterNot(_ == edge)
          adjacencyList.update(node, Edge(newNode, startIx, splitLength1) :: adjacencyList.getOrElse(node, Nil))
          adjacencyList(newNode) =
            List(Edge(nodeTo, startIx + splitLength1, splitLength2), Edge(newLeaf, nonMatchingIx, size - nonMatchingIx))
      }
    }

    private def findLastMatchingNode(adjacencyList: MutableMap[Int, List[Edge]],
                                     suffixStart: Int): (Int, Option[Edge], Int, Int) = {
      def findNextEdgeInPath(node: Int, letter: Char): Option[Edge] =
        adjacencyList.getOrElse(node, Nil).find{ case Edge(_, startIx, _) => text(startIx) == letter }

      @tailrec
      def loop(currentNode: Int, ix: Int): (Int, Option[Edge], Int, Int) =
        findNextEdgeInPath(currentNode, text(ix)) match {
          case None => (currentNode, None, ix, ix)
          case Some(edge @ Edge(nodeTo, _, length)) =>
            findFirstNonMatchingIndex(ix, edge) match {
              case Some(nonMatchingIx) => (currentNode, Some(edge), ix, nonMatchingIx)
              case None => loop(nodeTo, ix + length)
            }
        }

      loop(root, suffixStart)
    }

    private def findFirstNonMatchingIndex(ix: Int, edge: Edge): Option[Int] = {
      val Edge(_, startIx, length) = edge
      val substring: String = text.slice(startIx, startIx + length)
      substring
        .lazyZip(text.substring(ix))
        .lazyZip(ix until size)
        .find{ case (a, b, _) => a != b }
        .map{ case (_, _, jy) => jy }
    }
  }

  private def updateLongestRepeat(longestRepeat: String, nodes: Set[(Int, String)]): String =
    nodes.foldLeft(longestRepeat){ case (acc, (_, string)) => if (string.length > acc.length) string else acc }

  private def hasAtLeastTwoChildren(suffixTree: SuffixTree, node: Int): Boolean =
    suffixTree.adjacencyList.getOrElse(node, Nil).lengthCompare(2) >= 0

  private def getValidNeighbourNodes(suffixTree: SuffixTree, node: Int, spelledString: String): List[(Int, String)] =
    suffixTree
      .adjacencyList
      .getOrElse(node, Nil)
      .withFilter{ case Edge(nodeTo, _, _) => hasAtLeastTwoChildren(suffixTree, nodeTo) }
      .map{ case Edge(nodeTo, startIx, length) =>
        (nodeTo, spelledString + suffixTree.text.slice(startIx, startIx + length))
      }

  def findLongestRepeatInText(suffixTree: SuffixTree): String = {
    @tailrec
    def loop(longestRepeat: String, currentNodes: Set[(Int, String)]): String = {
      if (currentNodes.isEmpty) longestRepeat
      else {
        val nextNodes: Set[(Int, String)] =
          currentNodes.flatMap{ case (node, spelledString) => getValidNeighbourNodes(suffixTree, node, spelledString) }
        loop(updateLongestRepeat(longestRepeat, nextNodes), nextNodes)
      }
    }

    val emptyString: String = ""
    loop(emptyString, Set((suffixTree.root, emptyString)))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val tree = new SuffixTree(text + '$')
    val result: String = findLongestRepeatInText(tree)
    println(result)
  }
}

File: Python/textbook_track/chapter09/ba9d.py
# Find the Longest Repeat in a String
import itertools as it
import sys

from collections import defaultdict, namedtuple

Edge = namedtuple('Edge', ['node_to', 'substring_start', 'substring_length'])


class SuffixTree:
    def __init__(self, text):
        self._text = text
        self._size = len(text)
        self._node_id_generator = it.count()
        self._root = next(self._node_id_generator)
        self._adjacency_list = self._build_suffix_tree()

    @property
    def text(self):
        return self._text

    @property
    def root(self):
        return self._root

    @property
    def adjacency_list(self):
        return self._adjacency_list

    def edges(self):
        for edge_labels in self._adjacency_list.values():
            for _, start_ix, length in edge_labels:
                yield self._text[start_ix:start_ix+length]

    def _build_suffix_tree(self):
        adjacency_list = defaultdict(list)
        for ix in range(self._size):
            self._add_suffix_to_tree(adjacency_list, ix)
        return dict(adjacency_list)

    def _add_suffix_to_tree(self, adjacency_list, suffix_start):
        node, edge, edge_start_ix, non_matching_ix = self._find_last_matching_node(adjacency_list, suffix_start)
        new_leaf = next(self._node_id_generator)
        if edge is None:
            adjacency_list[node].append(Edge(new_leaf, edge_start_ix, self._size - edge_start_ix))
        else:
            new_node = next(self._node_id_generator)
            split_length_1 = non_matching_ix - edge_start_ix
            split_length_2 = edge.substring_length - split_length_1
            adjacency_list[node].remove(edge)
            adjacency_list[node].append(Edge(new_node, edge.substring_start, split_length_1))
            adjacency_list[new_node].append(Edge(edge.node_to, edge.substring_start+split_length_1, split_length_2))
            new_edge = Edge(new_leaf, non_matching_ix, self._size - non_matching_ix)
            adjacency_list[new_node].append(new_edge)

    def _find_last_matching_node(self, adjacency_list, suffix_start):
        ix = suffix_start
        current_node = self._root
        while (edge := self._find_next_edge_in_path(adjacency_list, current_node, self._text[ix])) is not None:
            non_matching_ix = self._find_first_non_matching_index(ix, edge)
            if non_matching_ix is not None:
                return current_node, edge, ix, non_matching_ix

            current_node = edge.node_to
            ix += edge.substring_length

        return current_node, None, ix, ix

    def _find_next_edge_in_path(self, adjacency_list, node, letter):
        edges = adjacency_list.get(node, [])
        for edge in edges:
            if self._text[edge.substring_start] == letter:
                return edge
        return None

    def _find_first_non_matching_index(self, ix, edge):
        substring = self._text[edge.substring_start:edge.substring_start+edge.substring_length]
        for jy, (a, b) in enumerate(zip(self._text[ix:], substring), start=ix):
            if a != b:
                return jy
        return None


def has_at_least_two_children(suffix_tree, node):
    return len(suffix_tree.adjacency_list.get(node, [])) >= 2


def get_valid_neighbour_nodes(suffix_tree, node, spelled_string):
    valid_neighbours = []
    edges = suffix_tree.adjacency_list.get(node, [])
    for neighbour, substring_start, substring_length in edges:
        if has_at_least_two_children(suffix_tree, neighbour):
            edge_label = suffix_tree.text[substring_start:substring_start+substring_length]
            valid_neighbours.append((neighbour, spelled_string + edge_label))
    return valid_neighbours


def update_longest_repeat(longest_repeat, nodes):
    length = len(longest_repeat)
    for _, string in nodes:
        if len(string) > length:
            longest_repeat = string
            length = len(longest_repeat)

    return longest_repeat


def find_longest_repeat_in_text(suffix_tree):
    longest_repeat = ''
    current_nodes = {(suffix_tree.root, longest_repeat)}
    while current_nodes:
        next_nodes = set()
        for node, spelled_string in current_nodes:
            next_nodes.update(get_valid_neighbour_nodes(suffix_tree, node, spelled_string))

        longest_repeat = update_longest_repeat(longest_repeat, next_nodes)
        current_nodes = next_nodes

    return longest_repeat


def main():
    reader = sys.stdin
    text = next(reader).rstrip()
    tree = SuffixTree(text + '$')
    result = find_longest_repeat_in_text(tree)
    print(result)


if __name__ == '__main__':
    main()
